File: 백준/Gold/2294. 동전 2/동전 2.cc
#include <iostream>
#define INF 987654321
using namespace std;

int n, k;
int coin[101];
int DP[10001];

void DynamicPrograming() {

        for (int i = 1; i <=n; i++) {
            for (int j = coin[i]; j <= k; j++) {
                DP[j] = min(DP[j], DP[j - coin[i]]+1);
            }
        }

        if (DP[k] == INF) {
            cout << -1;
        }
        else {
            cout << DP[k];
        }
}

int main() {

	cin >> n >> k;
	for (int i = 1; i <= n; i++) {
		cin >> coin[i];
	}

	for (int i = 1; i <= k; i++) {
		DP[i] = INF;
	}

	DynamicPrograming();
}

File: 백준/Gold/2294. 동전 2/동전 2.java
import static java.lang.Math.min;

import java.security.spec.RSAOtherPrimeInfo;
import java.util.Scanner;

public class Main {
    private static final int INF = 987654321;
    private static int n, k;
    private static int[] coin = new int[101];
    private static int[] DP = new int[10001];

    private static void dynamicProgramming() {
        for (int i = 1 ; i <= n; i++) {
            for (int j = coin[i]; j <= k; j++) {
                DP[j] = min(DP[j], DP[j - coin[i]] + 1);
            }
        }

        if (DP[k] == INF) {
            System.out.println("-1");
            return;
        }
        System.out.println(DP[k]);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        k = scanner.nextInt();

        for (int i = 1; i <= n; i++) {
            coin[i] = scanner.nextInt();
        }

        for (int i = 1; i <= k; i++) {
            DP[i] = INF;
        }

        dynamicProgramming();
    }
}
